concent_api/concent_api/middleware.py

```diff
@@ -39,3 +50,80 @@ def __call__(self, request):
+    1) By default it HTML should not be used
+    2) If `text/html` is present in 'Accept' header and 'application/json' is not, HTML should be used
+    3) If both are present, HTML will be used if it has greater weight than JSON
+    """
+    def get_probability(selected_method, accepted_methods):
+        for method in accepted_methods:
+            if selected_method in method:
+                if ":q=" in method:
+                    return float(method.split(":q=")[1])
+        return 1
+
+    def get_html_probability():
+        if 'headers' in request.META and 'Accept' in request.META['headers'] and "text/html" in request.META['headers']['Accept']:
+            return get_probability("text/html", request.META['headers']['Accept'])
+        return 0
+
+    def get_json_probability():
+        if 'headers' in request.META and 'Accept' in request.META['headers'] and "application/json" in request.META['headers']['Accept']:
+            return get_probability("application/json", request.META['headers']['Accept'])
+        return 0
+
+    html_probability = get_html_probability()
+    json_probability = get_json_probability()
+    if html_probability > 0:
+        if json_probability > 0:
+            return html_probability > json_probability
+        return True
+    return False
+
+
+class HandleServerErrorMiddleware(object):
+    """
+    Used to catch all unhandled exceptions and return JSON response containing error information.
+    """
+    def __init__(self, get_response):
+        self.get_response = get_response
+        self.sid = None
+
+    def __call__(self, request):
+        self.sid = transaction.savepoint()
+        response = self.get_response(request)
+        transaction.savepoint_commit(self.sid)
+        return response
+
+    def process_exception(self, request, exception):  # pylint: disable=unused-argument
+        transaction.savepoint_rollback(self.sid)
+        logger.error(
+            'Internal Server Error: %s', request.path,
+            exc_info=sys.exc_info(),
+            extra={'status_code': 500, 'request': request},
+        )
+        debug_info_enabled = getattr(settings, 'DEBUG_INFO_IN_ERROR_RESPONSES', settings.DEBUG)
+        if should_use_html(request):
+            return self._build_html_response(debug_info_enabled)
+        return self._build_json_response(exception, debug_info_enabled)
+
+    @staticmethod
+    def _build_html_response(debug_info_enabled):
+        message = '<h1>Server Error (500)</h1>'
+        if debug_info_enabled:
+            message += f"\n\nSTACK TRACE:\n{traceback.format_exc()}"
+        return HttpResponseServerError(message, content_type='text/html')
+
+    @staticmethod
+    def _build_json_response(exception, debug_info_enabled):
+        message = {
+            "error_message": str(exception) or DEFAULT_ERROR_MESSAGE,
+            "error_code": getattr(exception, "error_code", ErrorCode.CONCENT_APPLICATION_CRASH.value),
```

Comment:
I remember that I pointed out to @rwrzesien in the pull request that added an exception with error code that we should use the enum items (rather than names or values) wherever possible. And I think that the `error_code` field contains an item of `ErrorCode` type rather than it's value. And if it does not, it probably should. Please check, because if so, you need to call `.value` on it too.

Reply:
done